<!-- <div style="border:solid green 2px; padding: 20px"> -->
<b>Привет, Андрей!</b>

Меня зовут Кирилл Юрин, и я буду проверять твой проект. Предлагаю общаться на «ты» 🙂 Но если это неудобно — дай знать, и мы перейдем на «вы». 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ниже ты найдешь мои комментарии — пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера</b></font>
    
Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера</b></font>
    
Так выделены дополнительные материалы и развивающие замечания. Их отработка остается на твое усмотрение.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера</b></font>
    
Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
    
Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
    
</div>

---

<div class="alert alert-success">
<font size="4"><b>Обратная связь v.1</b></font>
    
Ты выполнил проект на высоком уровне. Анализ выдержан в верном ключе.

Чтобы завершить проект, тебе нужно доработать подготовку данных и раздел со стат. тестами. Думаю, у тебя с этим не будет проблем, ведь уже подавляющая часть проекта выполнена корректно 🙂
    
Буду ждать работу вновь!</div>

---

<div class="alert alert-success">
<font size="4"><b>Обратная связь v.2</b></font>
    
Андрей, спасибо, что оперативно доработал проект! БОльшая часть критических замечаний тобой учтена, но поработай еще, пожалуйста, над разделом со стат. тестами. Уверен, что ты справишься.
    
Дорабатывай проект и присылай снова — буду ждать! </div>

---

# Проект "Анализ результатов AB-теста" #
Цель проекта: проверка изменений направленных на увелечение выручки интернет-магазина

# Содержание #
<div>
  <ol>
      <li>
          <a href="#data_description">Описание данных</a>
      </li>
      <li>
          <a href="#research_plan">План проведения исследования</a>
      </li>
      <li>
          <a href="#import">Импорт библиотек</a>
      </li>
      <li>
          <a href="#data">Загрузка данных</a>
      </li>
      <ul>
          <li>
              <a href="#preview_hypothesis">Предварительный осмотр hypothesis</a>
          </li>
          <li>
              <a href="#preview_orders">Предварительный осмотр orders</a>
          </li>
          <li>
              <a href="#preview_visitors">Предварительный осмотр visitors</a>
          </li>
      </ul>
      <li>
          <a href="#prepare">Предобработка данных</a>
      </li>
      <ul>
          <li>
              <a href="#prepare_data">Изменение данных</a>
          </li>
          <li>
              <a href="#prepare_type">Изменение типов данных</a>
          </li>
          <li>
              <a href="#prepare_dupl">Проверка наличия дубликатов</a>
          </li>
      </ul>
      <li>
          <a href="#hypothesis">Приоритезация гипотез</a>
      </li>
      <ul>
          <li>
              <a href="#hypothesis_ice">Фреймворк ICE, RICE</a>
          </li>
      </ul>
      <li>
          <a href="#abtest">Анализ A/B-теста</a>
      </li>
      <ul>
          <li>
              <a href="#abtest_cum_revenue">Кумулятивная выручка групп</a>
          </li>
          <li>
              <a href="#abtest_cum_reciept_mean">Кумулятивный средний чек групп</a>
          </li>
          <li>
              <a href="#abtest_cum_reciept_mean_diff">Относительное изменение кумулятивного среднего чека групп</a>
          </li>
          <li>
              <a href="#abtest_cum_conversion">Кумулятивная конверсия групп</a>
          </li>
          <li>
              <a href="#abtest_cum_conversion_diff">Относительное изменение кумулятивной конверсии групп</a>
          </li>
          <li>
              <a href="#abtest_orders_by_users">Распределение количества заказов пользователей</a>
          </li>
          <li>
              <a href="#abtest_revenue_by_orders">Распределение стоимости заказов пользователей</a>
          </li>
          <li>
              <a href="#abtest_stat_norm">Проверка нормальности распределения данных групп</a>
          </li>
          <ul>
              <li>
                  <a href="#abtest_stat_norm_conv_a">Проверка нормальности распределения кумулятивной конверсии группы А</a>
              </li>
              <li>
                  <a href="#abtest_stat_norm_conv_b">Проверка нормальности распределения кумулятивной конверсии группы B</a>
              </li>
              <li>
                  <a href="#abtest_stat_norm_reciept_a">Проверка нормальности распределения кумулятивной средней суммы чека группы A</a>
              </li>
              <li>
                  <a href="#abtest_stat_norm_reciept_b">Проверка нормальности распределения кумулятивной средней суммы чека группы B</a>
              </li>
          </ul>
          <li>
              <a href="#abtest_prepare_data">Подготовка данных для теста Уилкоксона-Манна-Уитни</a>
          </li>
          <ul>
              <li>
                  <a href="#abtest_abnormal_users">Обработка выбросов</a>
              </li>
              <ul>
                  <li>
                      <a href="#abtest_abnormal_users_99">Список пользователей 99-процентиля</a>
                  </li>
                  <li>
                      <a href="#abtest_abnormal_users_95">Список пользователей 95-процентиля</a>
                  </li>
                  <li>
                      <a href="#abtest_abnormal_users_99_del">Удаление пользователей 99-процентиля</a>
                  </li>
                  <li>
                      <a href="#abtest_abnormal_users_95_del">Удаление пользователей 95-процентиля</a>
                  </li>
              </ul>
          </ul>
          <li>
              <a href="#abtest_stat_conv">Анализ различия конверсии групп</a>
          </li>
          <ul>
              <li>
                  <a href="#abtest_stat_conv_raw">Анализ различия конверсии групп. Сырые данные</a>
              </li>
              <li>
                  <a href="#abtest_stat_conv_99">Анализ различия конверсии групп. 99% пользователей</a>
              </li>
              <li>
                  <a href="#abtest_stat_conv_95">Анализ различия конверсии групп. 95% пользователей</a>
              </li>
          </ul>
          <li>
              <a href="#abtest_stat_reciept">Анализ различия суммы чека групп</a>
          </li>
          <ul>
              <li>
                  <a href="#abtest_stat_reciept_raw">Анализ различия суммы чека групп. Сырые данные</a>
              </li>
              <li>
                  <a href="#abtest_stat_reciept_99">Анализ различия суммы чека групп. 99% пользователей</a>
              </li>
              <li>
                  <a href="#abtest_stat_reciept_95">Анализ различия суммы чека групп. 95% пользователей</a>
              </li>
          </ul>
      </ul>
      <li>
          <a href="#summary">Вывод</a>
      </li>
  </ol>
</div>


<a id='data_description'></a>
# Описание данных #
1. Файл "/datasets/hypothesis.csv":<br>
 Hypothesis — краткое описание гипотезы;<br> 
 Reach — охват пользователей по 10-балльной шкале;<br> 
 Impact — влияние на пользователей по 10-балльной шкале;<br> 
 Confidence — уверенность в гипотезе по 10-балльной шкале;<br>
 Efforts — затраты ресурсов на проверку гипотезы по 10-балльной шкале. Чем больше значение Efforts, тем дороже проверка гипотезы.<br>
2. Файл "/datasets/orders.csv":<br>
 transactionId — идентификатор заказа;<br>
 visitorId — идентификатор пользователя, совершившего заказ;<br>
 date — дата, когда был совершён заказ;<br>
 revenue — выручка заказа;<br>
 group — группа A/B-теста, в которую попал заказ.<br>
3. Файл "/datasets/visitors.csv":<br>
 date — дата;<br>
 group — группа A/B-теста;<br>
 visitors — количество пользователей в указанную дату в указанной группе A/B-теста<br>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Хорошо, что добавил описание данных, но сейчас его довольно тяжело прочесть. Можешь воспользоваться `<br>` для переноса строки.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Привет, Кирилл!
Готово    
</div>


<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Намного лучше.</div>

<a id='research_plan'></a>
# План проведения исследования #
1. загрузка данных
2. предобработка данных
3. приоритезация гипотез по ICE
4. приоритезация гипотез по RICE
5. A/B-тест
6. анализ результатов A/B-теста

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Молодец, что оформил введение в проект, оно позволяет понять, насколько ты в курсе проблемы и понимаешь что делаешь. Ты сделаешь введение лучше, если добавишь в него краткое описание проекта и сформулируешь цели более развернуто. Еще твой проект станет намного удобней в использовании, если ты добавишь интерактивное оглавление — с помощью гиперссылок или в плагине toc. Материал об этом есть в студенческом notion, для получения ссылки и доступа обратись к куратору.</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Отличное оглавление!</div>

<a id='import'></a>
# Импорт библиотек #

In [1]:
import pandas as pd
import numpy as np
import datetime as dt
from IPython.display import display
# from IPython.core.display_functions import display
import plotly.express as px
from scipy import stats as st
import warnings
import os

pd.set_option('display.max_colwidth', None)
warnings.filterwarnings('ignore')

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.1</b></font>
    
Правильно, что весь импорт ты проводишь в первой ячейке работы. Так твой коллега, запускающий работу, будет в курсе используемых в ней библиотек и сможет при необходимости быстро настроить окружение. Однако стоило проверить, что все модули, которые ты используешь, есть в Jupyter hub. Например, `IPython.core.display_functions` — нет. </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Исправил импорт.
Импорт из указанного пакета IDE пометила как устаревший. 
Поэтому ранее использовал актуальную библиотеку.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Спасибо, что исправил.</div>

<a id='data'></a>
# Загрузка данных #

In [2]:
path_1 = 'datasets/'
path_2 = '/datasets/'

if os.path.exists(path=path_1):
    hypothesis = pd.read_csv(filepath_or_buffer=path_1 + 'hypothesis.csv', sep=',')
    orders = pd.read_csv(filepath_or_buffer=path_1 + 'orders.csv', sep=',')
    visitors = pd.read_csv(filepath_or_buffer=path_1 + 'visitors.csv', sep=',')
elif os.path.exists(path=path_2):
    hypothesis = pd.read_csv(filepath_or_buffer=path_2 + 'hypothesis.csv', sep=',')
    orders = pd.read_csv(filepath_or_buffer=path_2 + 'orders.csv', sep=',')
    visitors = pd.read_csv(filepath_or_buffer=path_2 + 'visitors.csv', sep=',')
else: print('Данные отсутствуют. Проверьте путь к папкам с данными')

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.1</b></font>
    
Кажется, работа над проектов велась локально и пути к файлам были изменены на общедоступные не совсем верно. Сейчас я внесу изменения в твой код, но в следующий раз будь внимательней. Стоит писать код таким образом, что бы он запускался и в jupiter hub, и локально без внесения дополнительных правок. Этого можно добиться многими способами. Например, использовать конструкцию ``try-except`` — для путей файлов: ``try`` — пути на локальном компьютере, ``except``  пути на сервере. Или применить библиотеку ``os``:

    import os

    pth1 = '/folder_1/data.csv'
    pth2 = '/folder_2/data.csv'
    
    if os.path.exists(pth1):
        query_1 = pd.read_csv(pth1)
    elif os.path.exists(pth2):
        query_1 = pd.read_csv(pth2)
    else:
        print('Something is wrong)
    
</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Исправлено.
</div>


<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Класс! И на будущее пригодится.</div>

<a id='preview_hypothesis'></a>
## Предварительный осмотр hypothesis ##

In [3]:
display(hypothesis)

Hypothesis  \
0            Добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей   
1                                     Запустить собственную службу доставки, что сократит срок доставки заказов   
2  Добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
3           Изменить структура категорий, что увеличит конверсию, т.к. пользователи быстрее найдут нужный товар   
4                              Изменить цвет фона главной страницы, чтобы увеличить вовлеченность пользователей   
5                      Добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов   
6          Показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию   
7              Добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок   
8                                                       Запустить акцию, дающую скидку на товар в день рождения   

   Reach  Impact  Confidence  Efforts  
0      3      10           8        6  
1      2       5           4       10  
2      8       3           7        3  
3      8       3           3        8  
4      3       1           1        1  
5      3       2           2        3  
6      5       3           8        3  
7     10       7           8        5  
8      1       9           9        5

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Для полного отображения гипотез в таблицах советую использовать надстройку pd.set_option('display.max_colwidth', -1).</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Благодарю.
Добавил.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Отлично!</div>

In [4]:
hypothesis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Hypothesis  9 non-null      object
 1   Reach       9 non-null      int64 
 2   Impact      9 non-null      int64 
 3   Confidence  9 non-null      int64 
 4   Efforts     9 non-null      int64 
dtypes: int64(4), object(1)
memory usage: 488.0+ bytes


<a id='preview_orders'></a>
## Предварительный осмотр orders ##

In [5]:
display(orders)

transactionId   visitorId        date  revenue group
0        3667963787  3312258926  2019-08-15     1650     B
1        2804400009  3642806036  2019-08-15      730     B
2        2961555356  4069496402  2019-08-15      400     A
3        3797467345  1196621759  2019-08-15     9759     B
4        2282983706  2322279887  2019-08-15     2308     B
...             ...         ...         ...      ...   ...
1192     2662137336  3733762160  2019-08-14     6490     B
1193     2203539145   370388673  2019-08-14     3190     A
1194     1807773912   573423106  2019-08-14    10550     A
1195     1947021204  1614305549  2019-08-14      100     A
1196     3936777065  2108080724  2019-08-15   202740     B

[1197 rows x 5 columns]

In [6]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   transactionId  1197 non-null   int64 
 1   visitorId      1197 non-null   int64 
 2   date           1197 non-null   object
 3   revenue        1197 non-null   int64 
 4   group          1197 non-null   object
dtypes: int64(3), object(2)
memory usage: 46.9+ KB


## Предварительный осмотр visitors ##

In [7]:
display(visitors)

date group  visitors
0   2019-08-01     A       719
1   2019-08-02     A       619
2   2019-08-03     A       507
3   2019-08-04     A       717
4   2019-08-05     A       756
..         ...   ...       ...
57  2019-08-27     B       720
58  2019-08-28     B       654
59  2019-08-29     B       531
60  2019-08-30     B       490
61  2019-08-31     B       718

[62 rows x 3 columns]

In [8]:
visitors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62 entries, 0 to 61
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   date      62 non-null     object
 1   group     62 non-null     object
 2   visitors  62 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 1.6+ KB


## Результаты предварительного осмотра данных ##
1. hypothesis:
  - имена столбцов перевести в нижний регистр
  - hypothesis['hypothesis'] - перевести значение в нижний регистр
2. orders:
  - переименовать "transactionId" -> "transaction_id"
  - переименовать "visitorId" -> "visitor_id"
  - "date".type(object) - > "date".type(datetime)
3. visitors:
  - "date".type(object) -> "date".type(datetime)

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Ты верно определил особенности данных в части названия столбцов и несоответствия типов данных, хотя переводить текст гипотез в нижний регистр нет необходимости.</div>

<a id='prepare'></a>
# Предобработка данных #

<a id='prepare_rename_columns'></a>
## Переименоваине столбцов ##

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Молодец, что определяешь свою функцию, но не понятно почему не используешь ее далее.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Удалил.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Хозяин — барин 😉</div>

In [9]:
orders = orders.rename(columns={'transactionId':'transaction_id',
                                'visitorId':'visitor_id'})

hypothesis.columns = hypothesis.columns.str.lower().str.strip()
orders.columns = orders.columns.str.lower().str.strip()
visitors.columns = visitors.columns.str.lower().str.strip()

<a id='prepare_data'></a>
## Изменение данных ##

In [10]:
hypothesis['hypothesis'] = hypothesis['hypothesis'].str.strip().str.lower()

<a id='prepare_type'></a>
## Изменение типов данных ##

In [11]:
orders['date'] = orders['date'].map(lambda x: dt.datetime.strptime(x, '%Y-%m-%d'))
visitors['date'] = visitors['date'].map(lambda x: dt.datetime.strptime(x, '%Y-%m-%d'))

<a id='prepare_dupl'></a>
## Проверка наличия дубликатов ##

In [12]:
print(hypothesis.duplicated().sum())
print(orders.duplicated().sum())
print(visitors.duplicated().sum())

0
0
0


<a id='prepare_crossing'></a>
## Проверка пересечения групп ##

In [13]:
print(orders[orders['group'] == 'A'].isin(orders[orders['group'] == 'B']).sum())
print()
print(visitors[visitors['group'] == 'A'].isin(visitors[visitors['group'] == 'B']).sum())

transaction_id    0
visitor_id        0
date              0
revenue           0
group             0
dtype: int64

date        0
group       0
visitors    0
dtype: int64


## Итоги предобработки данных ##
1. Переименованы столбцы датафреймов:
  - orders:
    * 'transactionId' -> 'transaction_id'
    * 'visitorId' -> 'visitor_id'
  - имена всех столбцов всех датафреймов переведены в нижний регистр
2. Изменение значений:
  - hypothesis['hypothesis'] переведены в нижний регистр
3. Изменение типа данных:
  - orders['date'].dtype('object') -> orders['date'].dtype('datetime')
  - visotors['date'].dtype('object') -> visotors['date'].dtype('datetime')
4. Дубликаты в данных отсутствуют
5. Пересечения данных в группах отсутсвуют

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1</b></font>
    
Давай зафиксируем все промежуточные выводы по результатам предобработки данных. </div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Забыл упомянуть про изменение типа данных в `visitors`.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Поправил.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Обычно перед анализом результатов A/B-теста аналитики проверяют группы A и В на пересечение — здесь ты тоже можешь выполнить это упражнение. </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Расскажи пожалуйста подробнее о чем речь и я выполню упражнение.
Ну и риторический вопрос: что мешало команде готовившей курс рассказать об этом шаге и добавить его в план выполнения проекта?! :):):)
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Теоритически при наборе пользователей в группы для А/В-теста один и тот же пользователь может попасть как в группу А, так и в группу В, чего, конечно, не должно происходить, но иногда бывает. Как ты понимаешь, такое пересечение групп не может положительно сказаться на результатах А/В-теста. И тогда возникает вопрос: какова доля таких пользователей. Если доля не велика, то можно либо вычистить таких пользователей, либо оставить, т.к. мы понимаем, что их влияние также не велико. Но вот если доля пересекающихся пользователей значительна, то результатам такого А/В-теста будет мало доверия. В данном проекте у тебя есть возможность проверить пересечение групп только по тем пользователям, которые сделали заказы. По второму вопросу: каждый проект проверяет определенные навыки, и по мере твоего продвижения по курсу эти навыки будут расширяться. В текущем проекте это упражнение не является обязательным, чтобы не перегружать его, но в будущем ты уже будешь обращать на этот момент внимание 😉   </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Добавил
</div>

<a id='hypothesis'></a>
# Приоритезация гипотез #

<a id='hypothesis_ice'></a>
## Фреймворк ICE ##

In [14]:
hypothesis['ICE'] = (hypothesis['impact'] * hypothesis['confidence']) / hypothesis['efforts']
hypothesis['RICE'] = (hypothesis['reach'] * hypothesis['impact'] * hypothesis['confidence']) / hypothesis['efforts']
print(hypothesis[['hypothesis', 'ICE', 'RICE']].sort_values(by=['ICE', 'RICE'], ascending=False))

                                                                                                     hypothesis  \
8                                                       запустить акцию, дающую скидку на товар в день рождения   
0            добавить два новых канала привлечения трафика, что позволит привлекать на 30% больше пользователей   
7              добавить форму подписки на все основные страницы, чтобы собрать базу клиентов для email-рассылок   
6          показать на главной странице баннеры с актуальными акциями и распродажами, чтобы увеличить конверсию   
2  добавить блоки рекомендаций товаров на сайт интернет магазина, чтобы повысить конверсию и средний чек заказа   
1                                     запустить собственную службу доставки, что сократит срок доставки заказов   
5                      добавить страницу отзывов клиентов о магазине, что позволит увеличить количество заказов   
3           изменить структура категорий, что увеличит конверсию, т.к. пользоват

## Итоги приоритезации гипотез ##
1. различия в оценках гипотез ICE и RICE обусловлены использованием компоненты 'REACH'(охват) в фреймфорке RICE.
   В фреймфорке ICE указанный  компонент не учитывается.

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Верно, фреймворки отличаются учетом охвата. Но что можно сказать, например, о составе ТОП-5 гипотез по каждому из фреймворков? </div>

<a id='abtest'></a>
# Анализ A/B-теста #

<a id='abtest_cum_revenue'></a>
## Кумулятивная выручка групп ##

In [15]:
dates_groups = orders[['date', 'group']].drop_duplicates()

### Агрегирование данных о покупателях/заказах/выручке групп, на указанную дату ###

In [16]:
orders_agg = dates_groups.apply(lambda x: orders[np.logical_and(orders['date'] <= x['date'],
                                                                orders['group'] == x['group'])]\
                                .agg({'date':np.max,
                                      'group':np.max,
                                      'transaction_id':pd.Series.nunique,
                                      'visitor_id':pd.Series.nunique,
                                      'revenue':np.sum}), axis='columns')\
    .sort_values(by=list(dates_groups.columns))

### Агрегирование данных о количестве посещений групп, на указанную дату ###

In [17]:
visitors_agg = dates_groups.apply(lambda x: visitors[np.logical_and(visitors['date'] <= x['date'],
                                                                    visitors['group'] == x['group'])]\
                                  .agg({'date':np.max,
                                        'group':np.max,
                                        'visitors':np.sum}), axis='columns')\
    .sort_values(by=list(dates_groups.columns))

### Объединение агрегированных таблиц заказов и посещений для формирования таблицы кумулятивных данных ###

In [18]:
cum_data = orders_agg.merge(right=visitors_agg,
                            how='inner',
                            right_on=['date', 'group'],
                            left_on=['date', 'group'])
cum_data.columns = ['date', 'group', 'orders', 'buyers', 'revenue', 'visitors']

### График кумулятивной выручки групп ###

In [19]:
fig = px.line(data_frame=cum_data,
              x='date',
              y='revenue',
              color='group',
              title='Кумулятивная выручка групп',
              template='seaborn',
              labels={'revenue':'Выручка',
                      'date':'Дата',
                      'group':'Группа'})
fig.show()

## Итоги анализа кумулятивной выручки групп ##
1. кумулятивная выручка группв B растет быстрее кумулятивной выручки группы A
2. 19 августа наблюдается резкий рост кумулятивной выручки в группе B
   Возможные причины:
   - крупные покупки пользователей из группы B
   - большое количество заказов пользователей группы B

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Ты произвел корректные расчеты, сделал хорошую визуализацию и отметил в выводах скачек и вероятные причины, которые его повлекли.</div>

<a id='abtest_cum_reciept_mean'></a>
## Кумулятивный средний чек групп ##

In [20]:
fig = px.line(data_frame=cum_data,
              x='date',
              y=cum_data['revenue'] / cum_data['orders'],
              color='group',
              title='Кумулятивный средний чек групп',
              template='seaborn',
              labels={'y':'Средняя сумма чека',
                      'date':'Дата',
                      'group':'Группа'})
fig.show()

## Итоги анализа кумулятивного среднего чека групп ##
1. средний чек пользователей группы B выше чем в группе А
2. 19 августа наблюдается резкий рост среднего чека группы B
   Возможные причины:
   - крупные покупки пользователей из группы B 19 августа

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
И здесь визуализация с выводами на высоте.</div>

<a id='abtest_cum_reciept_mean_diff'></a>
## Относительное изменение кумулятивного среднего чека групп ##

In [21]:
cum_revenue_a = cum_data[cum_data['group'] == 'A'][['date', 'revenue', 'orders']]
cum_revenue_a['reciept_mean'] = cum_revenue_a['revenue'] / cum_revenue_a['orders']

cum_revenue_b = cum_data[cum_data['group'] == 'B'][['date', 'revenue', 'orders']]
cum_revenue_b['reciept_mean'] = cum_revenue_b['revenue'] / cum_revenue_b['orders']

merge_cum_revenue = cum_revenue_a.merge(right=cum_revenue_b,
                                        how='left',
                                        left_on='date',
                                        right_on='date',
                                        suffixes=['A','B'])
merge_cum_revenue['cum_reciept_diff'] = merge_cum_revenue['reciept_meanB'] / merge_cum_revenue['reciept_meanA'] - 1

### График относительного изменения значения кумулятивного среднего чека групп ###

In [22]:
fig = px.line(data_frame=merge_cum_revenue,
              x='date',
              y='cum_reciept_diff',
              template='seaborn',
              labels={'cum_reciept_diff':'Относительное значение кумулятивного среднего чека',
                      'date':'Дата'},
              title='Относительное изменение значения кумулятивного среднего чека группы B')
fig.add_shape(type='line',
              line_color='salmon',
              x0=merge_cum_revenue['date'].min(),
              x1=merge_cum_revenue['date'].max(),
              y0=0,
              y1=0,
              line_width=3)
fig.show()

## Итоги анализа графика относительного изменения кумулятивного среднего чека группы B ##
1. до 2 августа сумма среднего чека была ниже чем в группе А
2. со второго по 11 августа сумма среднего чека группы B превышала сумму среднего чека группы А.
   Пиковое значение превышения суммы чека, в указанный период зафиксировано 8 и 6 августа.
   Разница в суммах среднего чека 6 и 8 августа составила 35%.
3. В период с 11 по 15 августа сумма среднего чека группы была чем у группы А.
   Пиковое значение падения суммы среднего чека группы B зафиксировано 13 августа.
   Падение составило 13%.
4. 15 августа сумма средних чеков в обеих группах сравнялась.
   До 18 августа наблюдается медленный рост суммы среднего чека группы B.
5. 19 августа зафиксирован резкий рост суммы среднего чека группы B.
   Рост составил 49%.
6. Относительное изменение кумулятивной средней суммы чека не стабилизировалось.
6. Необходимо провести анализ выбросов средней суммы чека в обеих группах.

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.1</b></font>
    
Визуализация корректная, только поправлю, что мы анализируем в данном случае именно кумулятивный средний чек. Что можно сказать о том стабилизировались данные или нет? </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Исправил.
Добавил.
</div>

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.2</b></font>
    
Ты отмечаешь, что за время эксперимента относительное изменение кумулятивного среднего чека несколько раз сильно менялось — разве это говорит нам о том, что данные стабилизировались? А на графике не поправил заголовок и подпись оси Y. </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Исправил.
Добавил.
</div>

<a id='abtest_cum_conversion'></a>
## Кумулятивная конверсия групп ##

In [23]:
fig = px.line(data_frame=cum_data,
              x='date',
              y=cum_data['buyers'] / cum_data['visitors'],
              color='group',
              title='Кумулятивная конверсия групп',
              template='seaborn',
              labels={'y':'Кумулятивная конверсия',
                      'date':'Дата',
                      'group':'Группа'})
fig.show()

## Итоги анализа конверсии групп ##
1. до 18 августа наблюдаются колебания значения конверсии для обеих групп
2. 18 августа конверсия обеих групп выравнивается.
3. конверсия в группе B выше конверсии в группе А ~0.3%

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Отличная визуализация и правильные выводы.</div>

<a id='abtest_cum_conversion_diff'></a>
## Относительное изменение кумулятивной конверсии групп ##

In [24]:
cum_buyers_a = cum_data[cum_data['group'] == 'A'][['date', 'buyers', 'visitors']]
cum_buyers_a['conversion'] = cum_buyers_a['buyers'] / cum_buyers_a['visitors']

cum_buyers_b = cum_data[cum_data['group'] == 'B'][['date', 'buyers', 'visitors']]
cum_buyers_b['conversion'] = cum_buyers_b['buyers'] / cum_buyers_b['visitors']

merge_cum_conversion = cum_buyers_a.merge(right=cum_buyers_b,
                                        how='left',
                                        left_on='date',
                                        right_on='date',
                                        suffixes=['A','B'])
merge_cum_conversion['cum_conversion_diff'] = merge_cum_conversion['conversionB'] / merge_cum_conversion['conversionA'] - 1

In [25]:
fig = px.line(data_frame=merge_cum_conversion,
              x='date',
              y='cum_conversion_diff',
              template='seaborn',
              labels={'cum_conversion_diff':'Кумулятивная конверсия',
                      'date':'Дата'},
              title='Относительное изменение кумулятивной конверсии группы B')
fig.add_shape(type='line',
              line_color='salmon',
              x0=merge_cum_revenue['date'].min(),
              x1=merge_cum_revenue['date'].max(),
              y0=0,
              y1=0,
              line_width=3)
fig.show()

## Итоги анализа графика изменения кумулятивной конверсии группы B относительно конверсии группы А ##
1. на протяжении августа 2019 конверсия группы B выше конверсии группы A
2. исключение составляет период 03.08.2019 - 05.08.2019

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Повторюсь, хорошая визуализация, а анализируем мы в данном случае кумулятивную конверсию.

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Исправил.
Добавил.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Еще бы на графике поправить.

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Исправил.
</div>

<a id='abtest_orders_by_users'></a>
## Распределение количества заказов пользователей ##

In [26]:
orders_by_users = orders.groupby(by=['visitor_id'], as_index=False)\
    .agg({'transaction_id':pd.Series.nunique})
orders_by_users.columns = ['visitor_id', 'orders']

In [27]:
fig = px.scatter(data_frame=orders_by_users,
                 x=range(0, len(orders_by_users)),
                 y='orders',
                 template='seaborn',
                 title='Распределение количества заказов пользователей',
                 labels={'orders':'Кол-во заказов',
                         'x':'Кол-во пользователей'})
fig.show()

In [28]:
print(np.percentile(orders_by_users['orders'], [95, 99]))

[2. 4.]


## Итоги анализа распределения количества заказов пользователей ##
1. 5% пользователей сделали более 2 заказов
2. 1% пользователей сделали более 4 заказов
3. для дальнейшего исследования будем считать аномальными количество заказов пользователей > 4.

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.1</b></font>
    
Отличная визуализация и верно рассчитанные перцентили. Но не хватает выводов о наличии аномалий в данных.</div>

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.2</b></font>
    
Не увидел ниже вывода по наличию аномалий по количеству заказов на пользователя.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Добавил в "Итоги анализа распределения количества заказов пользователей"
</div>

<a id='abtest_revenue_by_orders'></a>
## Распределение стоимости заказов пользователей ##

In [29]:
fig = px.scatter(data_frame=orders,
                 x=range(0, len(orders['revenue'])),
                 y='revenue',
                 template='seaborn',
                 title='Распределение суммы заказов пользователей',
                 labels={'revenue':'Сумма заказа',
                         'x':'Кол-во заказов'})
fig.show()

In [30]:
print(np.percentile(orders['revenue'], [95,99]))

[28000.  58233.2]


## Итоги анализа распределения суммы заказов пользователей ##
1. 5% пользователей сделали заказ на сумму более 28 тыс. руб.
2. 1% пользователей сделали заказ на сумму более 58 тыс. руб.
3. для дальнейшего исследования будем считать аномальными заказы сумма которых превышает 58 тыс. руб.

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>
    
Хорошая визуализация и корректные расчеты перцентилей. Аномальные значения ты определил — можно переходить к проверке гипотез.</div>

<a id='abtest_stat_norm'></a>
## Проверка нормальности распределения данных групп ##

<a id='abtest_stat_norm_conv_a'></a>
### Проверка нормальности распределения кумулятивной конверсии группы А ###
$H_0:$ кумулятивная конверсия группы А распределена нормально<br>
$H_1:$ кумулятивная конверсия группы А распределена ненормально<br>
Уровень статистической значимости: 0.05<br>

In [31]:
alpha = 0.05  # критический уровень статистической значимости

results = st.shapiro(merge_cum_conversion['conversionA'])
p_value = results.pvalue

print(f'p-значение: {p_value :0.5%}')

if p_value < alpha:
    print('Отвергаем нулевую гипотезу: распределение не нормально')
else:
    print('Не получилось отвергнуть нулевую гипотезу, всё нормально')

p-значение: 0.00003%
Отвергаем нулевую гипотезу: распределение не нормально


<a id = 'abtest_stat_norm_conv_b'></a>
### Проверка нормальности распределения кумулятивной конверсии группы B ###
$H_0:$ кумулятивная конверсия группы B распределена нормально<br>
$H_1:$ кумулятивная конверсия группы B распределена ненормально<br>
Уровень статистической значимости: 0.05<br>

In [32]:

alpha = 0.05  # критический уровень статистической значимости

results = st.shapiro(merge_cum_conversion['conversionB'])
p_value = results.pvalue

print(f'p-значение: {p_value :0.5%}')

if p_value < alpha:
    print('Отвергаем нулевую гипотезу: распределение не нормально')
else:
    print('Не получилось отвергнуть нулевую гипотезу, всё нормально')

p-значение: 63.32517%
Не получилось отвергнуть нулевую гипотезу, всё нормально


<a id='abtest_stat_norm_reciept_a'></a>
### Проверка нормальности распределения кумулятивной средней суммы чека группы A ###
$H_0:$ кумулятивная средняя сумма чека группы А распределена нормально<br>
$H_1:$ кумулятивная средняя сумма чека группы А распределена ненормально<br>
Уровень статистической значимости: 0.05<br>

In [33]:
alpha = 0.05  # критический уровень статистической значимости

results = st.shapiro(merge_cum_revenue['revenueA'])
p_value = results.pvalue

print(f'p-значение: {p_value :0.5%}')

if p_value < alpha:
    print('Отвергаем нулевую гипотезу: распределение не нормально')
else:
    print('Не получилось отвергнуть нулевую гипотезу, всё нормально')

p-значение: 19.39247%
Не получилось отвергнуть нулевую гипотезу, всё нормально


<a id='abtest_stat_norm_reciept_b'></a>
### Проверка нормальности распределения кумулятивной средней суммы чека группы B ###
$H_0:$ кумулятивная средняя сумма чека группы B распределена нормально<br>
$H_1:$ кумулятивная средняя сумма чека группы B распределена ненормально<br>
Уровень статистической значимости: 0.05<br>

In [34]:
alpha = 0.05  # критический уровень статистической значимости

results = st.shapiro(merge_cum_revenue['revenueB'])
p_value = results.pvalue

print(f'p-значение: {p_value :0.5%}')

if p_value < alpha:
    print('Отвергаем нулевую гипотезу: распределение не нормально')
else:
    print('Не получилось отвергнуть нулевую гипотезу, всё нормально')

p-значение: 1.03858%
Отвергаем нулевую гипотезу: распределение не нормально


## Итоги проверки нормальности распределения данных групп ##
1. распределение кумулятивной конверсии группы А ненормально
2. распределение кумулятивной средней суммы чека группы B ненормально
3. для дальнейшего анализа различия конверсии групп необходимо использовать критерий Уилкоксона-Манна-Уитни
4. для дальнейшего анализа различия средней суммы чека групп необходимо использовать критерий Уилкоксона-Манна-Уитни

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Молодец, что делаешь проверку на нормальность!
</div>

<a id='abtest_prepare_data'></a>
## Подготовка данных для теста Уилкоксона-Манна-Уитни ##

### Агрегированные данные пользователей группы А ###

In [35]:
users_a = (orders[orders['group'] == 'A']
                     .groupby('visitor_id', as_index=False)
                     .agg({'transaction_id':pd.Series.nunique,
                           'revenue':np.sum})
                     )
users_a.columns = ['user_id', 'orders', 'revenue']
users_a['mean_reciept'] = users_a['revenue'] / users_a['orders']

### Агрегированные данные пользователей группы B ###

In [36]:
users_b = (orders[orders['group'] == 'B']
                     .groupby('visitor_id', as_index=False)
                     .agg({'transaction_id':pd.Series.nunique,
                           'revenue':np.sum})
                     )
users_b.columns = ['user_id', 'orders', 'revenue']
users_b['mean_reciept'] = users_b['revenue'] / users_b['orders']

### Формирование данных для теста Уилкоксона-Манна-Уитни. Группа А ###

In [37]:
sample_a = pd.concat(
    [
        users_a[['orders', 'revenue', 'mean_reciept']],
        pd.DataFrame(data = {'orders':0,
                             'revenue':0,
                             'mean_reciept':0},
                     index=np.arange(merge_cum_conversion['visitorsA'].sum() - len(users_a)))
    ],
    axis='index'
)

### Формирование данных для теста Уилкоксона-Манна-Уитни. Группа B ###

In [38]:
sample_b = pd.concat(
    [
        users_b[['orders', 'revenue', 'mean_reciept']],
        pd.DataFrame(data = {'orders':0,
                             'revenue':0,
                             'mean_reciept':0},
                     index=np.arange(merge_cum_conversion['visitorsB'].sum() - len(users_b)))
    ],
    axis='index'
)

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>

Ты делаешь верные расчеты. </div>

<a id='abtest_abnormal_users'></a>
### Обработка выбросов ###

<a id='abtest_abnormal_users_99'></a>
#### Список пользователей 99-процентиля ####

In [39]:
abnormal_users_99 = pd.concat(
    [
        users_a
        [
            np.logical_or(
                users_a['orders'] > np.percentile(orders_by_users['orders'], [99])[0],
                users_a['revenue'] > np.percentile(orders['revenue'], [99])[0]
                          )
        ],
        users_b
        [
            np.logical_or
            (
                users_b['orders'] > np.percentile(orders_by_users['orders'], [99])[0],
                users_b['revenue'] > np.percentile(orders['revenue'], [99])[0]
            )
        ]
    ],
    axis='index'
).drop_duplicates()

<div class="alert alert-warning">
<font size="4">📝 <b>Комментарий ревьюера v.1</b></font>

Лучше не прописывать значения перцентилей вручную — так тебе нужно будет каждый раз при изменении исходных данных править код. Гораздо удобнее использовать переменные или непосредственно вычислять перцентили. </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Исправлено.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Супер!
</div>

<a id='abtest_abnormal_users_95'></a>
#### Список пользователей 95-процентиля ####

In [40]:
abnormal_users_95 = pd.concat(
    [
        users_a
        [
            np.logical_or(
                users_a['orders'] > np.percentile(orders_by_users['orders'], [95])[0],
                users_a['revenue'] > np.percentile(orders['revenue'], [95])[0]
                          )
        ],
        users_b
        [
            np.logical_or
            (
                users_b['orders'] > np.percentile(orders_by_users['orders'], [95])[0],
                users_b['revenue'] > np.percentile(orders['revenue'], [95])[0]
            )
        ]
    ],
    axis='index'
).drop_duplicates()

<a id='abtest_abnormal_users_99_del'></a>
#### Удаление пользователей 99-процентиля ####

In [41]:
users_a_filtered_99 = pd.concat(
    [
        users_a[np.logical_not(users_a['user_id'].isin(abnormal_users_99['user_id']))][['orders', 'revenue', 'mean_reciept']],
        pd.DataFrame(data = {'orders':0,
                             'revenue':0,
                             'mean_reciept':0},
                     index=np.arange(merge_cum_conversion['visitorsA'].sum() - len(users_a)))
    ],
    axis='index'
)

users_b_filtered_99 = pd.concat(
    [
        users_b[np.logical_not(users_b['user_id'].isin(abnormal_users_99['user_id']))][['orders', 'revenue', 'mean_reciept']],
        pd.DataFrame(data = {'orders':0,
                             'revenue':0,
                             'mean_reciept':0},
                     index=np.arange(merge_cum_conversion['visitorsB'].sum() - len(users_b)))
    ],
    axis='index'
)

<a id='abtest_abnormal_users_95_del'></a>
#### Удаление пользователей 95-процентиля ####

In [42]:
users_a_filtered_95 = pd.concat(
    [
        users_a[np.logical_not(users_a['user_id'].isin(abnormal_users_95['user_id']))][['orders', 'revenue', 'mean_reciept']],
        pd.DataFrame(data = {'orders':0,
                             'revenue':0,
                             'mean_reciept':0},
                     index=np.arange(merge_cum_conversion['visitorsA'].sum() - len(users_a)))
    ],
    axis='index'
)

users_b_filtered_95 = pd.concat(
    [
        users_b[np.logical_not(users_b['user_id'].isin(abnormal_users_95['user_id']))][['orders', 'revenue', 'mean_reciept']],
        pd.DataFrame(data = {'orders':0,
                             'revenue':0,
                             'mean_reciept':0},
                     index=np.arange(merge_cum_conversion['visitorsB'].sum() - len(users_b)))
    ],
    axis='index'
)


<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1</b></font>

Перед проведением тестов важно сформулировать нулевую и альтернативную гипотезы. Ты можешь их сформулировать в общем виде или же конкретно для каждой проверки. Но важно, чтобы они были. Это позволяет произвести полный анализ: от гипотезы к выводам. И в рамках обучения мне важно увидеть, что у тебя с этим нет проблем. И не забудь зафиксировать уровень стат. значимости.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Добавлено ниже.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Увидел — отлично!
</div>

<a id='abtest_stat_conv'></a>
## Анализ различия конверсии групп ##
$H_0:$ конверсия групп одинакова<br>
$H_1:$ конверсия групп различается<br>
Уровень статистической значимости: 0.05<br>

<a id='abtest_stat_conv_raw'></a>
### Анализ различия конверсии групп. Сырые данные ###

In [43]:
alpha = 0.05
p_value = st.mannwhitneyu(sample_a["orders"], sample_b["orders"]).pvalue
print(f'p_value: {p_value :0.3%}')
if p_value < alpha:
    print('Отвергаем нулевую гипотезу: разница статистически значима')
else:
    print('Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя')

print(f'Относительное значение конверсии группы B: {sample_b["orders"].mean() / sample_a["orders"].mean() - 1 :.3%}')

p_value: 1.465%
Отвергаем нулевую гипотезу: разница статистически значима
Относительное значение конверсии группы B: 14.370%


<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1
    </b></font>

Ты делаешь верные расчеты, но здесь также важно зафиксировать промежуточные выводы по проведенному стат. тесту о том удалось ли тебе отвергнуть нулевую гипотезу или нет, есть ли между величинами стат. значимые различия.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Оформлено ниже в виде промежуточного вывода.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Через `if-else` тоже норм.
</div>

<a id='abtest_stat_conv_99'></a>
### Анализ различия конверсии групп. 99% пользователей ###

In [44]:
alpha = 0.05
p_value = st.mannwhitneyu(users_a_filtered_99["orders"], users_b_filtered_99["orders"]).pvalue
print(f'p_value: {p_value :0.3%}')
if p_value < alpha:
    print('Отвергаем нулевую гипотезу: разница статистически значима')
else:
    print('Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя')

print(f'Относительное значение конверсии группы B: {users_b_filtered_99["orders"].mean() / users_a_filtered_99["orders"].mean() - 1 :.3%}')

p_value: 1.038%
Отвергаем нулевую гипотезу: разница статистически значима
Относительное значение конверсии группы B: 16.475%


<a id='abtest_stat_conv_95'></a>
### Анализ различия конверсии групп. 95% пользователей ###

In [45]:
alpha = 0.05
p_value = st.mannwhitneyu(users_a_filtered_95["orders"], users_b_filtered_95["orders"]).pvalue
print(f'p_value: {p_value :0.3%}')
if p_value < alpha:
    print('Отвергаем нулевую гипотезу: разница статистически значима')
else:
    print('Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя')

print(f'Относительное значение конверсии группы B: {users_b_filtered_95["orders"].mean() / users_a_filtered_95["orders"].mean() - 1 :.3%}')

p_value: 0.927%
Отвергаем нулевую гипотезу: разница статистически значима
Относительное значение конверсии группы B: 18.803%


## Итоги анализа различия конверсии групп ##
1. Различие конверсии групп статистически значимо на сырых и очищенных данных.
   Рост средней конверсии группы B составил:
   - сырые данные: 14.37%
   - 99% пользоватлей: 16.475%
   - 95% пользователей: 18.803%

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.1</b></font>

Молодец, что пробуешь разные варианты фильтрации данных, хотя вполне достаточно было взять 95-й перцентиль.</div>

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1</b></font>

Не хватает формулировки выводов по результатам стат. тестов на очищенных данных.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Добавлено.
</div>

<div class="alert alert-success">
<font size="4">👍 <b>Комментарий ревьюера v.2</b></font>
    
Ты делаешь верные выводы.
</div>

<a id='abtest_stat_reciept'></a>
## Анализ различия суммы чека групп ##
$H_0:$ сумма чека групп одинакова<br>
$H_1:$ сумма чека групп различается<br>
Уровень статистической значимости: 0.05<br>

<a id='abtest_stat_reciept_raw'></a>
### Анализ различия суммы чека групп. Сырые данные ###

In [46]:
alpha = 0.05
revenue_a = orders[orders['group'] == 'A']["revenue"]
revenue_b = orders[orders['group'] == 'B']["revenue"]
p_value = st.mannwhitneyu(revenue_a, revenue_b).pvalue
print(f'p_value: {p_value :0.3%}')
if p_value < alpha:
    print('Отвергаем нулевую гипотезу: разница статистически значима')
else:
    print('Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя')

print(f'Относительное значение суммы чека группы B: {revenue_b.mean() / revenue_a.mean() - 1 :.3%}')

p_value: 72.929%
Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя
Относительное значение суммы чека группы B: 25.871%


<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1</b></font>

В данном случае речь идет о среднем чеке выборки заказов, а не о выборке из средних чеков пользователей. И добавь, пожалуйста, вывод.</div>

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.2</b></font>

Теперь ты использовал выборку из сумм покупок агрегированных по пользователям, но при таком подходе у тебя может очень сильно отличаться средний чек групп. Например, в одной группе будут пользователи с десятью покупками по рублю, а во второй с одной покупкой по 10 рублей. Для таких групп сумма покупок по пользователям будет одинаковой, а средний чек будет сильно отличаться. Чтобы сравнить средние чеки групп нужно использовать неагрегированные данные по суммам покупок. </div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Исправлено.
</div>

<a id='abtest_stat_reciept_99'></a>
### Анализ различия суммы чека групп. 99% пользователей ###

In [47]:
alpha = 0.05
revenue_a = orders[np.logical_and(orders['group'] == 'A',
                                  ~orders['visitor_id'].isin(abnormal_users_99['user_id']))]["revenue"]
revenue_b = orders[np.logical_and(orders['group'] == 'B',
                                  ~orders['visitor_id'].isin(abnormal_users_99['user_id']))]["revenue"]
p_value = st.mannwhitneyu(revenue_a, revenue_b).pvalue
print(f'p_value: {p_value :0.3%}')
if p_value < alpha:
    print('Отвергаем нулевую гипотезу: разница статистически значима')
else:
    print('Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя')

print(f'Относительное значение суммы чека группы B: {revenue_b.mean() / revenue_a.mean() - 1 :.3%}')

p_value: 70.539%
Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя
Относительное значение суммы чека группы B: 2.154%


<a id='abtest_stat_reciept_95'></a>
### Анализ различия суммы чека групп. 95% пользователей ###

In [48]:
alpha = 0.05
revenue_a = orders[np.logical_and(orders['group'] == 'A',
                                  ~orders['visitor_id'].isin(abnormal_users_95['user_id']))]["revenue"]
revenue_b = orders[np.logical_and(orders['group'] == 'B',
                                  ~orders['visitor_id'].isin(abnormal_users_95['user_id']))]["revenue"]
p_value = st.mannwhitneyu(revenue_a, revenue_b).pvalue
print(f'p_value: {p_value :0.3%}')
if p_value < alpha:
    print('Отвергаем нулевую гипотезу: разница статистически значима')
else:
    print('Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя')

print(f'Относительное значение суммы чека группы B: {revenue_b.mean() / revenue_a.mean() - 1 :.3%}')

p_value: 84.903%
Не получилось отвергнуть нулевую гипотезу, вывод о различии сделать нельзя
Относительное значение суммы чека группы B: -1.279%


## Итоги анализа различия суммы чеков групп ##
1. Различие в сумме чеков групп не имеет статистического значения.
   Рост средней суммы чека группы B составил:
   - сырые данные: 25.871%
   - 99% пользоватлей: 2.154%
   - 95% пользователей: -1.279%

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1</b></font>

Здесь тоже нужно использовать выборки стоимости заказов. И также не хватает промежуточных выводов.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Добавлено.
</div>

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.2</b></font>

Выводы на месте, но расчет нужно поправить.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Исправлено.
</div>

<a id='summary'></a>
# Вывод #
Основываясь на результатах проведенного исследования признать сделанные изменения:
 - увеличения конверсии: признать успешными.
   Относительный рост конверсии в группе B составил:
   - сырые данные: 14.37%
   - 99% пользоватлей: 16.475%
   - 95% пользователей: 18.803%
 - увеление средней суммы чека: признать безуспешными.
   Относительный рост средней суммы чека в группе B составил:
   - сырые данные: 25.871%
   - 99% пользоватлей: 2.154%
   - 95% пользователей: -1.279%

и остановить тестирование.

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.1</b></font>

Сформулируй, пожалуйста, промежуточные выводы по каждому проведенному тобой стат. тесту и затем переходи к принятию решения.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента</b></font>
Исправлено.
</div>

<div class="alert alert-danger">
<font size="4">❌ <b>Комментарий ревьюера v.2</b></font>

Дорабатываем промежуточные выводы.</div>

<div class="alert alert-info">
<font size="4"><b>Комментарий студента v.2</b></font>
Исправлено.
</div>